Archives/VolFile.cpp

```diff
@@ -298,50 +285,30 @@ namespace Archives
 		volWriter.Write(&padding, volInfo.paddedIndexTableLength - volInfo.indexTableLength);
 	}
 
-	bool VolFile::OpenAllInputFiles(CreateVolumeInfo &volInfo)
+	void VolFile::OpenAllInputFiles(CreateVolumeInfo &volInfo)
 	{
-		volInfo.stringTableLength = 0;
-
-		// Open all the input files and store indexing info
-		for (std::size_t i = 0; i < volInfo.fileCount(); i++)
-		{
-			HANDLE fileHandle = CreateFileA(
-				volInfo.filesToPack[i].c_str(),	// fileName
-				GENERIC_READ,			// access mode
-				0,						// share mode
-				nullptr,				// security attributes
-				OPEN_EXISTING,			// creation disposition
-				FILE_ATTRIBUTE_NORMAL,	// file attributes
-				nullptr);				// template
-			
-			volInfo.fileHandles.push_back(fileHandle);
-
-			if (fileHandle == INVALID_HANDLE_VALUE)// Check for errors
-			{
-				// Error opening file. Close already opened files and return error
-				for (i--; i >= 0; i--) {
-					CloseHandle(volInfo.fileHandles[i]);
-				}
-
-				return false;
+		for (const auto& filename : volInfo.filesToPack) {
+			try {
+				volInfo.fileStreamReaders.push_back(std::make_unique<FileStreamReader>(filename));
```

Comment:
Hmm, unrelated to the current changes, but I just noticed this and wanted to point it out.

There's a bit of a style issue here. Perhaps we should we clear the `fileStreamReaders` collection before the start of this loop. Clearing it does seem a bit silly and redundant, since it will be initialized to empty, and this function *should* only be called once.

I guess what we really want, and to be obvious in a verifiable way, is that the collection of open files corresponds with the list of file names. If this function was called twice, that wouldn't be the case.

What this really calls for is a [`map`](https://en.wikipedia.org/wiki/Map_(higher-order_function)), as exists in most functional programming languages. The concept is implemented in C++ as [`std::transform`](https://en.cppreference.com/w/cpp/algorithm/transform?), however, the syntax is a bit ugly. The examples in the documentation also don't make it look very compelling. If we mapped the filenames to FileStreamReader objects, it would clearly be correct without having to analyse outside of the function.

Anyway, no need to change anything now. If we refactor to use static methods for packing, it will likely touch or eliminate this issue then. Likewise, if we delay opening the files until the point they are being copied, it will also eliminate this issue. The issue being one of style and verifiability, not of correctness.

Reply:
In the long run I would like to see the write specific portion of VolumeInfo stripped out and use the VolumeInfo struct both when reading and writing volume files. This would probably also affect the implementation. For now I'll just clear the stream container in the function to be explicit.

Reply:
Agreed. The write specific portions for VOL creation are probably better as local variables.